pkg/pipeline/prepare_pipelines.go

```diff
@@ -225,22 +228,76 @@ func (ctx *pipelineContext) pipelineFileExists(pipelineFilePath string) error {
 	return err
 }
 
-func (ctx *pipelineContext) buildTasks(envName string, tasks []pipelinev1.Task, timestamp string) map[string]pipelinev1.Task {
+func (ctx *pipelineContext) buildTasks(envName string, tasks []pipelinev1.Task, timestamp string) (map[string]pipelinev1.Task, error) {
+	var errs []error
 	taskMap := make(map[string]pipelinev1.Task)
 	for _, task := range tasks {
 		originalTaskName := task.Name
 		taskName := fmt.Sprintf("radix-task-%s-%s-%s-%s", getShortName(envName), getShortName(originalTaskName), timestamp, ctx.hash)
-		task.ObjectMeta.Name = taskName
-		task.ObjectMeta.Annotations = map[string]string{defaults.PipelineTaskNameAnnotation: originalTaskName}
-		task.ObjectMeta.Labels = labels.GetLabelsForEnvironment(ctx, envName)
+		if task.ObjectMeta.Labels == nil {
+			task.ObjectMeta.Labels = map[string]string{}
+		}
+		if task.ObjectMeta.Annotations == nil {
+			task.ObjectMeta.Annotations = map[string]string{}
+		}
+
+		for k, v := range labels.GetLabelsForEnvironment(ctx, envName) {
+			task.ObjectMeta.Labels[k] = v
+		}
+
+		if val, ok := task.ObjectMeta.Labels[labels.AzureWorkloadIdenityUse]; ok && val == "true" {
+			validateAzureSkipContainers(&task)
+		}
+
+		task.ObjectMeta.Annotations[defaults.PipelineTaskNameAnnotation] = originalTaskName
 		if ctx.ownerReference != nil {
 			task.ObjectMeta.OwnerReferences = []metav1.OwnerReference{*ctx.ownerReference}
 		}
 		ensureCorrectSecureContext(&task)
+
+		task.ObjectMeta.Name = taskName
 		taskMap[originalTaskName] = task
 		log.Debugf("created the task %s", task.Name)
 	}
-	return taskMap
+	return taskMap, errors.Join(errs...)
+}
+
+func validateAzureSkipContainers(task *pipelinev1.Task) {
+	skip := strings.Split(task.ObjectMeta.Annotations[annotations.AzureWorkloadIdentiySkipContainers], ";")
+	var updatedSkipNames []string
+
+	for _, containerName := range skip {
+		// check for container name without prefix "step-", and directly(but then warn that step- should be added)
+		if slices.ContainsFunc(task.Spec.Steps, func(s pipelinev1.Step) bool {
+			return "step-"+s.Name == containerName
+		}) {
+			updatedSkipNames = append(updatedSkipNames, containerName)
+			continue
+		}
+
+		if slices.ContainsFunc(task.Spec.Steps, func(s pipelinev1.Step) bool {
+			return s.Name == containerName
+		}) {
+			log.Infof("Adding 'step-' prefix to ignored Azure Workload Identity containers in task '%s' for step '%s'", task.Name, containerName)
+			updatedSkipNames = append(updatedSkipNames, "step-"+containerName)
+			continue
+		}
+
+		// Task containers neither containerName or step-containerName,ignore it, but log the warning
+		log.Warnf("Trying to ignore unknown container '%s' in task '%s' for Azure Workload Identity", containerName, task.Name)
+	}
+
+	// Ignore Tekton init containers:
+	tektonInitContainers := []string{"place-scripts", "prepare"}
```

Comment:
These should always be included in updatedSkipNames. Init updatedSkipNames with these values